validation/src/header.rs

```diff
@@ -203,35 +200,46 @@ fn find_next_difficulty_in_chain(
     store: &impl HeaderStore,
     prev_header: &BlockHeader,
     prev_height: BlockHeight,
-) -> u32 {
+) -> (u32, u32) {
     // This is the maximum difficulty target for the network
     let pow_limit_bits = pow_limit_bits(network);
+    let mut headers_inspected = 0;
     match network {
         Network::Testnet | Network::Regtest => {
             let mut current_header = *prev_header;
             let mut current_height = prev_height;
-            let mut current_hash = prev_header.block_hash();
+            let mut current_hash = current_header.block_hash();
             let initial_header_hash = store.get_initial_hash();
 
             // Keep traversing the blockchain backwards from the recent block to initial
             // header hash.
-            while current_hash != initial_header_hash {
+            loop {
+                // Count headers inspected for testing purposes.
+                headers_inspected += 1;
+
+                // Check if non-limit PoW found or it's time to adjust difficulty.
                 if current_header.bits != pow_limit_bits
                     || current_height % DIFFICULTY_ADJUSTMENT_INTERVAL == 0
                 {
-                    return current_header.bits;
+                    return (current_header.bits, headers_inspected);
+                }
+
+                // Stop if we reach the initial header.
+                if current_hash == initial_header_hash {
+                    break;
                 }
 
-                // Traverse to the previous header
+                // Save the previous header's hash, so that there's no need to calculate it.
+                current_hash = current_header.prev_blockhash;
```

Comment:
nit: given that this is code that's only relevant for testnet/regtest, I wouldn't worry too much about optimizations and would prioritize readability instead. It seems that the only change we really need is changing line 230 in the previous version of the file from:

```
current_hash = current_header.prev_blockhash;
```

to:

```
current_hash = current_header.block_hash();
```

I'd personally find that logic more straightforward to read.

Reply:
Done, switched to using `block_hash()` for readability.

Reply:
Note: there are 2 changes
1) using `current_header.block_hash()` instead of `current_header.prev_blockhash` in order not to jump over one block
2) replacing while-condition to check the difficulty of the initial block and only then to stop.

But this (2) change breaks `test_is_header_valid_invalid_computed_target` test, because its  initial `header_705600` is actually having non-limit PoW that breaks an expectation of `ValidateHeaderError::InvalidPoWForComputedTarget`.

If my assumption about checking all the headers (including the initial one) is correct, then this `test_is_header_valid_invalid_computed_target` has to be modified. One option to modify it would be to have blocks with PoW limit, so it finds no target and fails with an expected error. WDYT?

Reply:
I've modified `test_is_header_valid_invalid_computed_target` to make it pass. PTAL.